File: Reliability Analysis Tools/4. Sequential Importance Sampling/SIS_MATLAB/SIS_MATLAB/main_example_6.m
%% Sequential importance sampling: Ex. 5 Ref. 2 - points outside of hypersphere
%{
---------------------------------------------------------------------------
Created by:
Felipe Uribe
Daniel Koutas
Engineering Risk Analysis Group   
Technische Universitat Munchen
www.bgu.tum.de/era
---------------------------------------------------------------------------
First version: 2018-05
---------------------------------------------------------------------------
Current version: 2022-04
* Inclusion of sensitivity analysis
---------------------------------------------------------------------------
Comments:
* The SIS method in combination with a Gaussian Mixture model can only be
  applied for low-dimensional problems, since its accuracy decreases
  dramatically in high dimensions.
---------------------------------------------------------------------------
Based on:
1."Sequential importance sampling for structural reliability analysis"
   Papaioannou et al.
   Structural Safety 62 (2016) 66-75
2."Bayesian inference of engineering models"
   Wolfgang Betz
   Ph.D. Thesis.
---------------------------------------------------------------------------
%}
clear; close all; clc;

rng(10);

%% definition of the random variables
d      = 2;          % number of dimensions
pi_pdf = repmat(ERADist('standardnormal','PAR'), d, 1);   % n independent rv

% % correlation matrix
% R = eye(d);   % independent case
% 
% % object with distribution information
% pi_pdf = ERANataf(pi_pdf,R);    % if you want to include dependence

%% limit state function
r = 5.26;    % radius
m = 1;       % m in [0,4]
g = @(u) 1 - (sqrt(sum(u.^2,2))/r).^2 - (u(:,1)/r).*((1-(sqrt(sum(u.^2,2))/r).^m)./(1+(sqrt(sum(u.^2,2))/r).^m));

%% Implementation of sensitivity analysis: 1 - perform, 0 - not perform
sensitivity_analysis = 1;

%% Samples return: 0 - none, 1 - final sample, 2 - all samples
samples_return = 2;

%% Sequential importance sampling
N      = 3000;    % total number of samples for each level
p      = 0.1;     % N/number of chains per level
k_init = 3;       % initial number of Gaussians in the Mixture Model (GM)
burn   = 0;       % burn-in period
tarCOV = 1.5;     % target COV of weights

fprintf('\nSIS method: \n');
method = 'aCS';
switch method
   case 'GM'
      [Pf_SIS, lv, samplesU, samplesX, k_fin, S_F1] = SIS_GM(N, p, g, pi_pdf, k_init, burn, tarCOV, sensitivity_analysis, samples_return);
   case 'aCS'
      [Pf_SIS, lv, samplesU, samplesX, S_F1] = SIS_aCS(N, p, g, pi_pdf, burn, tarCOV, sensitivity_analysis, samples_return);
   otherwise
      error('Choose GM or aCS methods');
end

%% Reference values
% The reference values for the first order indices
S_F1_ref   = [0.1857, 0.1857];

% Print reference values for the first order indices
fprintf("***Reference first order Sobol' indices: ***\n");
disp(S_F1_ref);

% reference solution
pf_ref = 1e-6;

% show p_f results
fprintf('\n***Reference Pf: %g ***', pf_ref);
fprintf('\n***SIS Pf: %g ***\n\n', Pf_SIS);

%% Plots
% plot samplesU
if ~isempty(samplesU{1})
    if d == 2
       xx    = -6:0.05:6; 
       nnp   = length(xx); 
       [X,Y] = meshgrid(xx);
       xnod  = cat(2,reshape(X',nnp^2,1),reshape(Y',nnp^2,1));
       Z     = g(xnod); 
       Z     = reshape(Z,nnp,nnp);
       figure; hold on;
       contour(X,Y,Z,[0,0],'r','LineWidth',3);  % LSF
       for j = 1:length(samplesU)
          u_j_samples= samplesU{j};
          plot(u_j_samples(:,1),u_j_samples(:,2),'.');
       end
       axis equal tight;
    end
end

File: Reliability Analysis Tools/1. First-Order Reliability Method/FORM_Python/FORM_Python/main_example_6.py
import numpy as np
import scipy as sp
import matplotlib.pylab as plt
from ERANataf import ERANataf
from ERADist import ERADist
from FORM_HLRF import FORM_HLRF
from FORM_fmincon import FORM_fmincon
plt.close('all')
"""
---------------------------------------------------------------------------
FORM using HLRF algorithm and fmincon: Ex. 5 Ref. 3 - points outside of hypersphere
---------------------------------------------------------------------------
Created by:
Matthias Willer
Felipe Uribe
Luca Sardi
Daniel Koutas
Engineering Risk Analysis Group
Technische Universitat Munchen
www.era.bgu.tum.de
---------------------------------------------------------------------------
Version 2022-04
* Inclusion of sensitivity analysis
---------------------------------------------------------------------------
Based on:
1."Structural reliability under combined random load sequences."
   Rackwitz, R., and B. Fiessler (1979).
   Computers and Structures, 9.5, pp 489-494
2."Lecture Notes in Structural Reliability"
   Straub (2016)
3. "Sequential importance sampling for structural reliability analysis"
   Papaioannou et al.
   Structural Safety 62 (2016) 66-75
---------------------------------------------------------------------------
"""

#  definition of the random variables
d      = 2          # number of dimensions
pi_pdf = list()
for i in range(d):
    pi_pdf.append(ERADist('standardnormal', 'PAR', np.nan)) # n independent rv

# correlation matrix
R = np.eye(d)   # independent case

# object with distribution information
pi_pdf = ERANataf(pi_pdf, R)    # if you want to include dependence

#  limit state function
r = 5.26
m = 1
#g = lambda u: 1 - (np.linalg.norm(u)/r)**2 - (u[0,:]/r)*((1-(np.linalg.norm(u)/r)**m)/(1+(np.linalg.norm(u)/r)**m))
g  = lambda u: 1 - (np.sqrt(np.sum(u**2))/r)**2 - (u[0]/r)*((1-(np.sqrt(np.sum(u**2))/r)**m)/(1+(np.sqrt(np.sum(u**2))/r)**m))
dg = lambda u: np.array([(((u[0]**2 + u[1]**2)**(1/2)/r)**m - 1)/(r*(((u[0]**2 + u[1]**2)**(1/2)/r)**m + 1)) - (2*u[0])/r**2 + (m*u[0]**2*((u[0]**2 + u[1]**2)**(1/2)/r)**(m - 1))/(r**2*(u[0]**2 + u[1]**2)**(1/2)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m + 1)) - (m*u[0]**2*((u[0]**2 + u[1]**2)**(1/2)/r)**(m - 1)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m - 1))/(r**2*(u[0]**2 + u[1]**2)**(1/2)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m + 1)**2), \
                          (m*u[0]*u[1]*((u[0]**2 + u[1]**2)**(1/2)/r)**(m - 1))/(r**2*(u[0]**2 + u[1]**2)**(1/2)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m + 1)) - (2*u[1])/r**2 - (m*u[0]*u[1]*((u[0]**2 + u[1]**2)**(1/2)/r)**(m - 1)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m - 1))/(r**2*(u[0]**2 + u[1]**2)**(1/2)*(((u[0]**2 + u[1]**2)**(1/2)/r)**m + 1)**2) ])

#  Implementation of sensitivity analysis: 1 - perform, 0 - not perform
sensitivity_analysis = 1

#  Solve the optimization problem of the First Order Reliability Method

# OPC 1. FORM using Hasofer-Lind-Rackwitz-Fiessler algorithm HLRF (Ref.1 Pag.128)
[u_star_hlrf, x_star_hlrf, beta_hlrf, Pf_hlrf, S_F1_hlrf, S_F1_T_hlrf] = FORM_HLRF(g, dg, pi_pdf, sensitivity_analysis)

# OPC 2. FORM using Python scipy.optimize.minimize()
[u_star_fmc, x_star_fmc, beta_fmc, Pf_fmc, S_F1_fmc, S_F1_T_fmc] = FORM_fmincon(g, [], pi_pdf,sensitivity_analysis)

# OPC 3. FORM using Python scipy.optimize.minimize() with analytical gradients
#[u_star_fmc, x_star_fmc, beta_fmc, Pf_fmc, S_F1_fmc, S_F1_T_fmc] = FORM_fmincon(g, dg, pi_pdf,sensitivity_analysis)

# Reference values
# The reference values for the first order indices
S_F1_ref   = [0.1857, 0.1857]

# Print reference values for the first order indices
print("\n***Reference first order Sobol' indices: ***\n", S_F1_ref)

# reference solution
pf_ref = 1e-6

# show p_f results
print('***Reference Pf: ', pf_ref, ' ***')
print('***FORM HLRF Pf: ', Pf_hlrf, ' ***')
print('***FORM fmincon Pf: ', Pf_fmc, ' ***\n')

#  Plot HLRF results
if d == 2:
    # grid points
    xx      = np.linspace(-6,6,100)
    [X1,X2] = np.meshgrid(xx,xx)
    xnod = np.array([np.reshape(X1,len(xx)**2),np.reshape(X2,len(xx)**2)])
    ZX = np.array([g(xnod[:,i]) for i in range(len(xx)**2)])
    ZX = np.reshape(ZX,[len(xx),len(xx)])

    # figure
    plt.figure()
    plt.axes().set_aspect('equal', 'box')
    plt.pcolor(X1,X2,ZX)
    plt.contour(X1,X2,ZX,[0])
    plt.plot(0,0,'k.')                                               # origin in standard
    plt.plot([0, u_star_hlrf[0]],[0, u_star_hlrf[1]],label='HLRF')   # reliability index beta
    plt.plot(u_star_hlrf[0],u_star_hlrf[1],'r*')                     # design point in standard
    plt.plot([0, u_star_fmc[0]],[0, u_star_fmc[1]],label='SLSQP')    # reliability index beta
    plt.plot(u_star_fmc[0],u_star_fmc[1],'bo')                       # design point in standard
    plt.title('Standard space')
    plt.legend()
    plt.show()

# END
